Update vtk writer and add utilities

marinlauber opened the pull request:
This updates the `vtk` writer and solver #102. 

I added some tests for the `vtkUtils.jl`, but I am not sure about the correct procedure when creating `temp` file in test @b-fg ?

I also have made an example of how the simulation can be restarted from a `.pvd` file.

b-fg commented:
First, note that you need to update this branch with master. We have already merged 1.9 extensions, so the vtk code is now in `ext/`.

marinlauber commented:
I also merged #105 into the pull request.

marinlauber commented on ext/WaterLilyWriteVTKExt.jl:
```diff
@@ -7,32 +7,33 @@ else
-        !isdir(dir) && mkdir(dir)
-        new(fname,dir,paraview_collection(fname),attrib,[0])
-    end
+struct VTKWriter
```
@b-fg I had to rename the `struct` and remove the constructor function since it gave me an error when I had an empty function called `vtkWriter` in `WaterLily.jl` and I was defining a `struct` here with the same name.

b-fg replied:
Yes, that is all good. Also in general types (and `struct`s) start with a capital letter.

marinlauber commented:
Before merging I want to run the test on a GPU.

marinlauber commented:
This now works on the GPU and can be run in the `example` folder. I have updated the example to start the sim on the CPU, save it to `vtk`, and restart it on the GPU.

I have updated the old examples to work with the new `extensions` approach.

b-fg commented:
Hey Marin, this looks great, good job :)
I will leave minor comments as review for you to consider.
Cheers!

b-fg commented on examples/TwoD_circle_restart.jl:
```diff
@@ -0,0 +1,42 @@
+
+# re-start the sim from a paraview file but on the GPU this time
+import CUDA
+@assert CUDA.functional()
```
this is already asserted by the CUDAExt module when importing CUDA together with WaterLily. No need to call again here.
```suggestion
```

b-fg commented on examples/TwoD_circle_vtk.jl:
What's the point on having two separate examples with VTK and VTK+restart? Can't we just have one?

marinlauber replied:
I can put them into one example, that's true.

b-fg commented on ext/WaterLilyWriteVTKExt.jl:
```diff
@@ -7,32 +7,33 @@ else
-    vtk_grid(name;attrib,T)
+    vtkWriter(fname;attrib,dir,T)
 
 Generates a `vtkWriter` that hold the collection name to which the `vtk` files are written.
```
```suggestion
`VTKWriter` type that holds the `collection` name to which the `vtk` files are written.
```

b-fg commented on ext/WaterLilyWriteVTKExt.jl:
Please update docs from `vtkWriter` to `VTKWriter` everywhere in this file.

b-fg commented on src/WaterLily.jl:
```diff
@@ -105,19 +105,23 @@ end
-function vtkWriter end
-function write! end
+# WriteVTKExt export
+include("../ext/WaterLilyDefaults.jl")
```
I am not a huge fan of this. I think we can have all in here. Otherwise following this methodology, we will need to add at least 2 files in `ext/` every time we set up a new package extension.

marinlauber replied:
I was thinking of having all the default empty functions in `ext/WaterLilyDefault.jl` and only adding a single new file if we add a new extension. But I will revert to the previous state; we can always do that later.

b-fg commented on ext/WaterLilyWriteVTKExt.jl:
```diff
@@ -7,68 +7,75 @@ else
+    pvd_collection(fname;append=false)
 
-Generates a `vtkWriter` that hold the collection name to which the `vtk` files are written.
+wrapper for a paraview_collection that returns a pvd file header
```
```suggestion
Wrapper for a `paraview_collection` that returns a pvd file header
```

b-fg commented on ext/WaterLilyWriteVTKExt.jl:
```diff
@@ -7,68 +7,75 @@ else
 
-return a `Dict` containing the name and bound funtion for the default attributes.
-The name is used as the key in the `vtk` file and the function generates the data
+return a `Dict` containing the `name`` and `bound_function` for the default attributes. 
```
```suggestion
Returns a `Dict` containing the `name` and `bound_function` for the default attributes. 
```

b-fg reviewed (approved):
Looks all good to me!
